python/ql/src/semmle/python/pointsto/PointsTo.qll

```diff
@@ -82,503 +21,310 @@ library class CfgOrigin extends @py_object {
         or
-        import_points_to(f, value, origin) and cls = theModuleType() and context.appliesToScope(f.getScope())
+        name_lookup_points_to_maybe_origin(f, context, value, origin_or_obj)
         or
-        attribute_load_points_to(f, context, value, cls, origin)
+        not exists(fast_local_variable(f)) and not exists(name_local_variable(f)) and
+        global_lookup_points_to_maybe_origin(f, context, value, origin_or_obj)
+    }
+
+    /** Holds if `var` refers to `(value, origin)` given the context `context`. */
+    pragma [noinline]
+    cached predicate variablePointsTo(EssaVariable var, PointsToContext context, ObjectInternal value, CfgOrigin origin) {
+        ssa_definition_points_to(var.getDefinition(), context, value, origin)
         or
-        exists(CfgOrigin orig |
-            getattr_points_to(f, context, value, cls, orig) and
-            origin = orig.asCfgNodeOrHere(f)
+        exists(EssaVariable prev |
+            ssaShortCut(prev, var) and
+            variablePointsTo(prev, context, value, origin)
         )
+    }
+
+    private predicate ssaShortCut(EssaVariable start, EssaVariable end) {
+        end.getDefinition().(PhiFunction).getShortCircuitInput() = start
         or
-        if_exp_points_to(f, context, value, cls, origin)
-        or
-        from_import_points_to(f, context, value, cls, origin)
-        or
-        use_points_to(f, context, value, cls, origin)
-        or
-        def_points_to(f, context, value, cls, origin)
-        or
-        Calls::call_points_to(f, context, value, cls, origin)
-        or
-        subscript_points_to(f, context, value, cls, origin)
-        or
-        sys_version_info_slice(f, context, cls) and value = theSysVersionInfoTuple() and origin = f
-        or
-        sys_version_info_index(f, context, value, cls) and origin = f
-        or
-        sys_version_string_char0(f, context, value, cls) and origin = f
-        or
-        six_metaclass_points_to(f, context, value, cls, origin)
-        or
-        binary_expr_points_to(f, context, value, cls, origin)
-        or
-        compare_expr_points_to(f, context, value, cls, origin)
+        /* Attribute assignments have no effect as far as value tracking is concerned, except for `__class__`. */
+        exists(AttributeAssignment def |
+            not def.getName() = "__class__" and
+            start = def.getInput() and
+            end.getDefinition() = def
+        )
         or
-        not_points_to(f, context, value, cls, origin)
+        /* Ignore the effects of calls on their arguments. PointsTo is an approximation, 
+         * but attempting to improve accuracy would be very expensive for very little gain. */
+        exists(ArgumentRefinement def |
+            start = def.getInput() and
+            end.getDefinition() = def
+        )
         or
-        value.(SuperCall).instantiation(context, f) and f = origin and cls = theSuperType()
+        exists(EssaVariable mid |
```

Comment:
I doubt it would make any difference, because it is in a recursive layer.
It wouldn't save any memory in the caller and we'd need to compute the transitive closure as well.